CIF-1206 - [Github #88] Connector caches MUST include the store view header value

cjelger opened the pull request:
## Related Issues

Fixes #88 and fixes #87 

## How Has This Been Tested?

Extended unit tests and tested locally.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes and the overall coverage did not decrease.
- [x] All unit tests pass on CircleCi.
- [x] I ran all tests locally and they pass.

cjelger committed:
CIF-1206 - [Github #88] Connector caches MUST include the store view header value

- add store view to product cache keys

cjelger committed:
CIF-1206 - [Github #88] Connector caches MUST include the store view header value

- fix NPE when caching is disabled

herzog31 commented on bundles/cif-connector-graphql/src/main/java/com/adobe/cq/commerce/graphql/magento/GraphqlDataServiceImpl.java:
```diff
@@ -159,13 +157,14 @@ public ProductInterface getProductBySku(String sku, String storeView) {
 
         try {
-            return productCache.get(sku, () -> getProductBySkuImpl(sku, storeView)).orElse(null);
+            String key = sku + StringUtils.defaultString(storeView, MAGENTO_DEFAULT_STORE);
```
It would be better to use a separator when generating the key. Otherwise duplicate keys because of overlaps in the sku and storeview might occur.

cjelger replied:
I generated all cache keys with the same method so it's kind of unified and now used a separator that is less likely (than `-`) to appear in skus and store views but ... this doesn't prevent collisions, it just makes them less likely to occur. Consider the following cases:

```
   SKU        STORE      SEPARATOR     CACHE KEY

sku|        store|view      ''       sku|store|view
sku|store   |view           ''       sku|store|view
sku         store|view      '|'      sku|store|view
sku|store   view            '|'      sku|store|view
```
They all generate the same keys, so using a separator doesn't really help. However I think the `|` character wouldn't be common in skus and store views, so I think we can consider the collision probability very low.

cjelger replied:
@herzog31 I finally decided to replace the string keys with a dedicated class to use an array as a key, so we completely remove the issue with strings.

cjelger committed:
CIF-1206 - [Github #88] Connector caches MUST include the store view header value

- feedback from PR: unify cache keys and use a common less likely to be used in skus and store views

cjelger committed:
CIF-1206 - [Github #88] Connector caches MUST include the store view header value

- replaced cache keys with a dedicated class for array keys

cjelger committed:
Merge branch 'master' into CIF-1206

# Conflicts:
#	bundles/cif-connector-graphql/src/main/java/com/adobe/cq/commerce/graphql/magento/GraphqlDataServiceImpl.java
